cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go

```diff
@@ -239,21 +253,38 @@ func (m *asgCache) DeleteInstances(instances []*AwsInstanceRef) error {
 	}
 
 	for _, instance := range instances {
-		params := &autoscaling.TerminateInstanceInAutoScalingGroupInput{
-			InstanceId:                     aws.String(instance.Name),
-			ShouldDecrementDesiredCapacity: aws.Bool(true),
-		}
-		resp, err := m.service.TerminateInstanceInAutoScalingGroup(params)
-		if err != nil {
-			return err
+		// check if the instance is a placeholder - a requested instance that was never created by the node group
+		// if it is, just decrease the size of the node group, as there's no specific instance we can remove
+		matched, err := regexp.MatchString(fmt.Sprintf("^%s\\d+$", placeholderInstanceNamePrefix), instance.Name)
+		if err == nil && matched {
+			klog.V(4).Infof("instance %s is detected as a placeholder, decreasing ASG requested size instead "+
+				"of deleting instance", instance.Name)
+			m.decreaseAsgSizeByOneNoLock(commonAsg)
+			// mark this instance using its name as a timeouted placeholder
+			asg := m.instanceToAsg[*instance]
+			delete(m.instanceToAsg, *instance)
+			instance.Name = TimeoutedPlaceholderName
```

Comment:
I seems very much not right. 

IMO the instance should be marked as "timed-out" in some routine which refreshes node groups state. I guess in the same place when you are creating placeholders you should also monitor if some of the place holders are "old enough" to treat them as timed out. Then you can mark those as such and when instances are listed set `OutOfResources` error for them.

Then the DeleteInstances will be called by the core logic for timed-out instances. And here you will only remove the place holder, and decrease the asg size.

Reply:
I get your point, but the problem is again in how AWS cloud provider works. Inside, it keeps instance IDs only and these IDs need to stay constant. It also doesn't track any timeouts - that's on the core's side, specifically in [clusterstate.go](https://github.com/kubernetes/autoscaler/blob/2543e3e5ea02867da4b40fca88976a9c575380ad/cluster-autoscaler/clusterstate/clusterstate.go#L288). The logic that tracks instances and handles registration timeouts is there. 

We either have to reuse/link this logic over here: [auto_scaling_groups.go](https://github.com/kubernetes/autoscaler/blob/2543e3e5ea02867da4b40fca88976a9c575380ad/cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go#L306) or assume that if a `DeleteInstances` call comes for a placeholder, it's always because of timeout related to non-existing ScaleUp request. Basically, this whole logic is to guard against just a single use case: when CAS is started when already `required > running` for an ASG and the instances never come up.

I think there are the following possible cases for placeholder instances:
1) ScaleUp request times out. In that case, we don't have to modify Nodes state at all, as the time out handler will scale down the related ASG, mark it as "backed off" and decrease its size, which will make placeholders to be gone on the next refresh (that starts every main loop).
2) As I wrote above, on CAS start "required > running" already; in that case, `DeleteNodes` will be invoked by the Core logic (after detecting, that the node is listed by the cloud provider, but not registered in k8s) and we assume, that node is deleted because of time out. One possible modification is to explicitly pass "isTimeout" flag here.

What do you think?

Reply:
> I get your point, but the problem is again in how AWS cloud provider works. Inside, it keeps instance IDs only and these IDs need to stay constant. It also doesn't track any timeouts - that's on the core's side, specifically in [clusterstate.go](https://github.com/kubernetes/autoscaler/blob/2543e3e5ea02867da4b40fca88976a9c575380ad/cluster-autoscaler/clusterstate/clusterstate.go#L288). The logic that tracks instances and handles registration timeouts is there.
> 
> We either have to reuse/link this logic over here: [auto_scaling_groups.go](https://github.com/kubernetes/autoscaler/blob/2543e3e5ea02867da4b40fca88976a9c575380ad/cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go#L306) or assume that if a `DeleteInstances` call comes for a placeholder, it's always because of timeout related to non-existing ScaleUp request. Basically, this whole logic is to guard against just a single use case: when CAS is started when already `required > running` for an ASG and the instances never come up.

My understanding was that you wanted speed up the reaction time of CA when node insntances coming from spot-backed ASG are not showing up.
If this is just about fixing bug that the 15-minutes timeout mechanism built-in into CA does not trigger up situation is simpler (but also benefits are smaller).

> 1. ScaleUp request times out. In that case, we don't have to modify Nodes state at all, as the time out handler will scale down the related ASG, mark it as "backed off" and decrease its size, which will make placeholders to be gone on the next refresh (that starts every main loop).

Yes. Should work that way.

> 2. As I wrote above, on CAS start "required > running" already; in that case, `DeleteNodes` will be invoked by the Core logic (after detecting, that the node is listed by the cloud provider, but not registered in k8s) and we assume, that node is deleted because of time out. One possible modification is to explicitly pass "isTimeout" flag here.

If we just added placeholders and not modify code in any other way we would get to stable state after DeleteNodes are called. Just ASG would not be backed-off this time. Is that a problem? Do we care about (possibly rare) case when CA is restarted while scale-up is in progress? The state will get stable eventually. I do not believe we want to optimize for restart scenario. CA was never designed to be restarted very often. It survives restarts but it does not necessarily behaves optimally in such situations.

Reply:
Is that 15 minute timeout the one that's configured by the `--max-node-provision-time` command-line flag? There are several flags with default values of 15 minutes, and there's a const `MaxNodeStartupTime` with the same value. It doesn't look like that one can be changed without rebuilding the program.

15 minutes is too long for us to wait to react to an ASG's failure to acquire a machine, so we'd like to be able to dial this down.

Reply:
@losipiuk OK, I'll try to split this patch in 2: first provide a general fix, that has the worst case of recovering in `2 * NodeStartupTime`, then create another one, that can work in just `1 * NodeStartupTime`

@seh It's `--max-node-provision-time`. And yeah... and the general recommendation is "do not change that". Still, in my tests, I dialed it down to 2 minutes and it seemed to work. YMMV.